### PR TITLE
[context] Layer context creation scaffolding

### PR DESCRIPTION
**Pull request by kparichay**
Added todo list for the layer context creation
Added minor new interfaces for the layer context
Disable in-place optimization and inference memory optimization

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/layers/layer_node.h [hunk 1]
```diff
@@ -286,7 +288,10 @@ class LayerNode : public ml::train::Layer, public GraphNode {
 
   RunLayerContext
     run_context; /**< context required for running/execution of the layer. This
-                    will also contain the properties of the layer. */
+                    will also contain the properties of the layer. The
+                    properties will be copied upon final creation. Editing
+                    properties of the layer after init will not the properties
+                    in the context/graph unless intended. */
```
Comment (zhoonit) [inline nntrainer/layers/layer_node.h:294]: I wonder about the purpose of this behavior. My concern is that we could save to the context with `std::any` but we can't save the type.

Comment (kparichay) [reply nntrainer/layers/layer_node.h:294]: to avoid updating properties once the layer is initialized.

Comment (kparichay) [reply nntrainer/layers/layer_node.h:294]: > My concern is that we could save to the context with std::any but we can't save the type.

you want to directly save to run_context?

Comment (zhoonit) [reply nntrainer/layers/layer_node.h:294]: How about layer node control this behavior or let it be free for now?

Comment (kparichay) [reply nntrainer/layers/layer_node.h:294]: let's keep it free for now. how to handle stopping the setProperties() after init() can be handled later.

Comment (zhoonit) [reply nntrainer/layers/layer_node.h:294]: Cool

Comment (zhoonit) [reply nntrainer/layers/layer_node.h:294]: > you want to directly save to run_context?

I mean properties have it's own type. So there is no way to have a member if we are not using std::any. Let's delay the discussion for now :)

Comment (kparichay) [reply nntrainer/layers/layer_node.h:294]: yes.

**Review by zhoonit — APPROVED** [review]
LGTM

**Review by lhs8928 — APPROVED** [review]
LGTM

**Comment by kparichay** [comment]
> Android build failing, please check and label `PR/OK2GO`

done

**Pull request merged**